core/math/random.cpp

```diff
@@ -95,6 +95,107 @@ Variant Random::choice(const Variant &p_from) {
 	return Variant();
 }
 
+Variant Random::choices(const Variant &p_from, int count, const Array &p_weights, bool is_cumulative) {
+	int sum = 0;
+	Array cumulative_weights;
+	Array weights;
+	Array indices;
+	Array weighted_choices;
+
+	if((p_from.get_type() == Variant::DICTIONARY) && p_weights.empty()){
+		Dictionary dict = p_from;
+		weights = dict.values();
+	} else {
+		weights = p_weights;
+	}
+
+	if(!weights.empty()){
+		if(is_cumulative) {
+			sum = weights.back();
+			cumulative_weights = weights;
+		} else {
+			for(int i = 0; i < weights.size(); i++) {
+				if(int(weights[i]) < 0 || weights[i].get_type() != Variant::INT) {
+					ERR_FAIL_V_MSG(Array(), "Weights must be positive integers.");
+				} else {
+					sum += (int)weights[i];
+					cumulative_weights.push_back(sum);
+				}
+			}
+		}
+
+		for(int i = 0; i < count; i++) {
+			int left = 0;
+			int right = weights.size();
+			int random_number = randi() % sum;
+			// bisect
+			while (left < right)
+			{
+				int mid = (left + right) / 2;
+				if(int(cumulative_weights[mid]) < random_number) {
+					left = mid + 1;
+				} else {
+					right = mid;
+				}
+			}
+			indices.push_back(left);
+		}
+	}
+
+	switch (p_from.get_type()) {
+		case Variant::STRING: {
+			String str = p_from;
+			ERR_FAIL_COND_V_MSG(str.empty(), Variant(), "String is empty.");
+			if(weights.empty()){
+				for(int i=0; i < count; i++) {
+					weighted_choices.push_back(str.substr((randi() % str.size()), 1));
+				}
+			} else {
+				ERR_FAIL_COND_V_MSG(str.size() != weights.size(), Variant(), "Size of weights does not match.");
```

Comment:
Note that `size()` and `length()` are different in `String`. `size()` includes null character, while `length` doesn't. Not sure whether this logic matters here though. But if it does, you'd likely need to use `str.length()` instead.